omod/src/test/java/org/openmrs/module/owa/web/controller/OwaRestControllerTest.java

```diff
@@ -135,11 +137,33 @@ public void install_wrongDownloadUrl() throws Exception {
 		HttpServletResponse response = new MockHttpServletResponse();
 		String downloadUrl = "https://bintray.com/openmrs/owa/download_file?file_path=notAZip.zip";
 		OwaRestController controller = (OwaRestController) applicationContext.getBean("owaRestController");
-		InstallAppRequestObject requestData = new InstallAppRequestObject(downloadUrl, "Cohort Builder OWA");
-		List<App> appList = controller.install(requestData, request, response);
+		InstallAppRequestObject requestData = new InstallAppRequestObject(downloadUrl);
+		ResponseEntity<List<App>> appList = controller.install(requestData, request, response);
 		Assert.assertEquals("owa.not_a_zip", request.getSession().getAttribute(WebConstants.OPENMRS_ERROR_ATTR));
 	}
-
+	
+	@Test
+	public void install_missingFilePath() throws Exception {
+		HttpServletRequest request = new MockHttpServletRequest(new MockServletContext(), "POST", "/rest/owa/installapp");
+		HttpServletResponse response = new MockHttpServletResponse();
+		String downloadUrl = "https://bintray.com/openmrs/owa/download_file";
+		OwaRestController controller = (OwaRestController) applicationContext.getBean("owaRestController");
+		InstallAppRequestObject requestData = new InstallAppRequestObject(downloadUrl);
+		ResponseEntity<List<App>> appList = controller.install(requestData, request, response);
+		Assert.assertEquals("Invalid URL", request.getSession().getAttribute(WebConstants.OPENMRS_ERROR_ATTR));
+	}
+	
+	@Test
+	public void install_urlEndsWithFileName() throws Exception {
+		HttpServletRequest request = new MockHttpServletRequest(new MockServletContext(), "POST", "/rest/owa/installapp");
+		HttpServletResponse response = new MockHttpServletResponse();
+		String downloadUrl = "https://dl.bintray.com/openmrs/owa/cohortbuilder-1.0.0-beta.zip";
```

Comment:
Will this test pass if one compiles the module without an internet connection?

Reply:
No but I have tried using mockito to mock the class and found out it does not work for static methods. The other option of setting up powermock to do the mocking for the static method also failed. This is also referenced to the issue I raised on [talk](https://talk.openmrs.org/t/testing-with-mockito-on-openmrs/14664).

Reply:
Those were mocking the data as it is being passed to the request object for example using `MockMultipartFile`. In my case the request has been sent and have to mock a specific function that is called during the handling of the request.